tests/test_tfe.py

```diff
@@ -195,27 +246,24 @@ def circuit_nn(dummy1, array, dummy2):
         grad_target = (np.array(1.), np.array([[0.5,  0.43879, 0], [0, -0.43879, 0]]), np.array(-0.4))
         cost_target = 1.03257
 
-        for circuit in [circuit_n1s, circuit_n1v, circuit_nn]:
-
-            args = (tfe.Variable(0.46), tfe.Variable([[2., 3., 0.3], [7., 4., 2.1]]), tfe.Variable(-0.13))
+        args = (tfe.Variable(0.46), tfe.Variable([[2., 3., 0.3], [7., 4., 2.1]]), tfe.Variable(-0.13))
 
-            def cost(x, array, y):
-                c = tf.cast(circuit(tf.constant(0.111), array, tf.constant(4.5)), tf.float32)
-                if c.shape != tf.TensorShape([]):
-                    c = c[0]  # get a scalar
-                return c +0.5*array[0,0] +x -0.4*y
+        def cost(x, array, y):
+            c = tf.cast(circuit(tf.constant(0.111), array, tf.constant(4.5)), tf.float32)
+            if c.shape != tf.TensorShape([]):
+                c = c[0]  # get a scalar
+            return c +0.5*array[0,0] +x -0.4*y
 
-            with tf.GradientTape() as tape:
-                cost_res = cost(*args)
-                grad_res = np.array([i.numpy() for i in tape.gradient(cost_res, [args[0], args[2]])])
+        with tf.GradientTape() as tape:
+            cost_res = cost(*args)
+            grad_res = np.array([i.numpy() for i in tape.gradient(cost_res, [args[0], args[2]])])
 
-            self.assertAllAlmostEqual(cost_res.numpy(), cost_target, delta=self.tol)
-            self.assertAllAlmostEqual(grad_res, np.fromiter(grad_target[::2], dtype=np.float32), delta=self.tol)
+        assert np.allclose(cost_res.numpy(), cost_target, atol=tol, rtol=0)
+        assert np.allclose(grad_res, np.fromiter(grad_target[::2], dtype=np.float32), atol=tol, rtol=0)
```

Comment:
These three tests bother me. It's not clear why the circuit/cost functions give the same target result for the same inputs. Since they are now split, and it doesn't matter so much whether they give different or same results, perhaps we can change the arguments and expected values to be different in all three tests?

Also, @josh146, these tests look like you run them first to determine the target answer

Reply:
I think it's quite obvious why the three circuits give the same results: they are just the same on wire 0. 

The objective of this test is not to check if the results are correctly calculated, but that the interoperability of the different return types works. In that vein, I think we could just keep the existing ones (because they work). It would be better to have something less obfuscated, but I think we can just carry on with that.

Reply:
ok then you have a better eye than me :laughing:. I stared at them for 15 minutes and couldn't easily figure out what they were doing. 

I'm all for keeping tests that are working, and testing for something that needs to be tested. What we should steer away from is having tests where it is not obvious or clear what they are actually testing

Reply:
I think the implementation of those tests could be streamlined, but that would probably cost more than half an hour.

Reply:
As a stop gap, perhaps best to include a one to two line comment within the test function, along the lines of https://github.com/XanaduAI/pennylane/pull/266#discussion_r306337869

Reply:
I think that is a good solution for now!

Reply:
I also applied it to test_torch.py